# Segmenting and Clustering Neighbourhoods in Toronto¶




Importing libraries



In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

In [2]:
import pandas
import requests
from bs4 import BeautifulSoup
website_text = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(website_text,'xml')

table = soup.find('table')
table_rows = table.find_all('tr')


table_contents=[]
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['Postal Code'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)



printing table contents

In [3]:
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...



The new Wiki link already has nieghborhood merged according to Postal Code and Borough


In [4]:
len(df['Postal Code'].unique())

103

In [5]:
df[df['Borough'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

In [6]:
df.shape

(103, 3)

# Part two of assignment

Installing Geocoder

In [7]:
pip install geocoder

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 98 kB 10.2 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [8]:
import geocoder

#### **Instead using the CSV to import data**

In [9]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

checking the dtypes just to be sure

In [10]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [11]:
df.dtypes

Postal Code     object
Borough         object
Neighborhood    object
dtype: object

Confirming shapes for both

In [12]:
df.shape


(103, 3)

In [13]:
data.shape

(103, 3)

The below code will join the 2 datas based on the Postal Code column

In [14]:
df = df.join(data.set_index('Postal Code'), on='Postal Code')
df



Postal Code                Borough  \
0           M3A             North York   
1           M4A             North York   
2           M5A       Downtown Toronto   
3           M6A             North York   
4           M7A           Queen's Park   
..          ...                    ...   
98          M8X              Etobicoke   
99          M4Y       Downtown Toronto   
100         M7Y  East Toronto Business   
101         M8Y              Etobicoke   
102         M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

Dropping index column to make it as close to the table shown in the assignment page

In [15]:
df = df.reset_index()

In [16]:
df.drop(['index'], axis = 'columns', inplace = True)


In [17]:
df.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                       Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                     Don Mills North  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

In [18]:
df = df.rename(index = {'level_0' : 'index'})

In [19]:
df.index.name = 'index'


In [20]:
df.head()

Postal Code           Borough                      Neighborhood  \
index                                                                   
0             M3A        North York                         Parkwoods   
1             M4A        North York                  Victoria Village   
2             M5A  Downtown Toronto         Regent Park, Harbourfront   
3             M6A        North York  Lawrence Manor, Lawrence Heights   
4             M7A      Queen's Park     Ontario Provincial Government   

        Latitude  Longitude  
index                        
0      43.753259 -79.329656  
1      43.725882 -79.315572  
2      43.654260 -79.360636  
3      43.718518 -79.464763  
4      43.662301 -79.389494

In [21]:
df.shape

(103, 5)

The number of rows are 103 even after join and the column has an additional 2 columns making a total of 5 columns, so we have the right data now

# Part 3 of assignment

#### Clustering Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [32]:
from geopy.geocoders import Nominatim 

In [33]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [34]:

!pip install folium

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes


Creating visualizations for the map of Toronto

In [37]:
import folium
# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

Initiating the Foursquare API

In [38]:


CLIENT_ID = 'AGHGCCBFEW1K5KOBLVNNDLME3WJCBJBEEH4D2L24Y2GNCHXF' # your Foursquare ID
CLIENT_SECRET = 'X3TCFLP35J0C3ZZXXVYM2FP5OWXB1HDLRXPWS31UASMTAEM2' # your Foursquare Secret
VERSION = '20180323' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)



Your credentails:
CLIENT_ID: AGHGCCBFEW1K5KOBLVNNDLME3WJCBJBEEH4D2L24Y2GNCHXF
CLIENT_SECRET:X3TCFLP35J0C3ZZXXVYM2FP5OWXB1HDLRXPWS31UASMTAEM2


Creating a function to get all the venue categories 

In [39]:

neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))



Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


In [40]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url



'https://api.foursquare.com/v2/venues/explore?&client_id=AGHGCCBFEW1K5KOBLVNNDLME3WJCBJBEEH4D2L24Y2GNCHXF&client_secret=X3TCFLP35J0C3ZZXXVYM2FP5OWXB1HDLRXPWS31UASMTAEM2&v=20180323&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [41]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6067126504959b5a1fbfb8dc'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 3,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'contact': {},
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
    

In [42]:


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']



In [43]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0             Brookbanks Park               Park  43.751976 -79.332140
1               Variety Store  Food & Drink Shop  43.751974 -79.333114
2  TTC stop - 44 Valley Woods           Bus Stop  43.755402 -79.333741

In [44]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Gather the venues in Toronto for each neighbourhood

In [45]:


toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )



Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [46]:
print(toronto_venues.shape)
toronto_venues.head()

(2112, 7)


Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1               Variety Store       43.751974       -79.333114   
2  TTC stop - 44 Valley Woods       43.755402       -79.333741   
3      Victoria Village Arena       43.723481       -79.315635   
4                   Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2               Bus Stop  
3           Hockey Arena  
4  Portuguese Restaurant

In [47]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                            
Agincourt                                                            4   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     22   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   25   
...                                                                ...   
Willowdale West                                                      5   
Willowdale, Newtonbrook                                              1   
Woburn                                                               4   
Woodbine Heights                                                     5   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighbourhood                                                             
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      22   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    25   
...                                                                 ...   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woburn                                                                4   
Woodbine Heights                                                      5   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     22              22   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   25              25   
...                                                ...             ...   
Willowdale West                                      5               5   
Willowdale, Newtonbrook                              1               1   
Woburn                                               4               4   
Woodbine Heights                                     5               5   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               22   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             25   
...                                                          ...   
Willowdale West                                                5   
Willowdale, Newtonbrook                                        1   
Woburn                                                         4   
Woodbine Heights                                               5   
York Mills West                                                2   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     4  
Alderwood, Long Branch                                        8 

## One Hot encoding the venue Categories

In [48]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Vegetarian / Vegan Restaurant  \
0                    0             0  ...                              0   
1                    0             0  ...                              0   
2                    0             0  ...                              0   
3                    0             0  ...                              0   
4                    0             0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 273 columns]

In [49]:


toronto_onehot.shape



(2112, 273)

In [50]:


toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()



Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0                 0.00           0.0  ...   
1               0.0                 0.00           0.0  ...   
2               0.0                 0.00           0.0  ...   
3               0.0                 0.00           0.0  ...   
4               0.0                 0.04           0.0  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 273 columns]

In [51]:


toronto_grouped.shape



(100, 273)

In [52]:


def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]



In [53]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1          Skating Rink                   Pharmacy           Coffee Shop   
2                  Bank                  Pet Store   Fried Chicken Joint   
3                  Bank        Japanese Restaurant    Chinese Restaurant   
4           Coffee Shop             Sandwich Place              Boutique   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Dumpling Restaurant   Distribution Center               Dog Run   
1                   Pub        Sandwich Place                   Gym   
2         Shopping Mall        Sandwich Place                 Diner   
3           Yoga Studio   Distribution Center               Dog Run   
4       Thai Restaurant              Pharmacy           Pizza Place   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                 Donut Shop              Drugstore  
1      Airport Terminal         Dim Sum Restaurant     Falafel Restaurant  
2           Supermarket  Middle Eastern Restaurant       Sushi Restaurant  
3      Doner Restaurant                 Donut Shop              Drugstore  
4                   Pub                 Restaurant                   Café

In [54]:


# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]



array([1, 0, 1, 1, 1, 1, 1, 1, 1, 4], dtype=int32)

In [55]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                      Neighborhood  \
index                                                                   
0             M3A        North York                         Parkwoods   
1             M4A        North York                  Victoria Village   
2             M5A  Downtown Toronto         Regent Park, Harbourfront   
3             M6A        North York  Lawrence Manor, Lawrence Heights   
4             M7A      Queen's Park     Ontario Provincial Government   

        Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
index                                                               
0      43.753259 -79.329656             4.0              Bus Stop   
1      43.725882 -79.315572             0.0           Pizza Place   
2      43.654260 -79.360636             1.0           Coffee Shop   
3      43.718518 -79.464763             1.0        Clothing Store   
4      43.662301 -79.389494             1.0           Coffee Shop   

      2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
index                                                                      
0                      Park      Food & Drink Shop           Yoga Studio   
1               Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                      Park                 Bakery                   Pub   
3         Accessories Store  Vietnamese Restaurant             Gift Shop   
4          Sushi Restaurant            Yoga Studio   Japanese Restaurant   

      5th Most Common Venue        6th Most Common Venue  \
index                                                      
0                 Drugstore          Distribution Center   
1         Electronics Store  Eastern European Restaurant   
2                   Theater               Breakfast Spot   
3               Coffee Shop       Furniture / Home Store   
4                      Café                        Diner   

      7th Most Common Venue       8th Most Common Venue 9th Most Common Venue  \
index                                                                           
0                   Dog Run            Doner Restaurant            Donut Shop   
1               Escape Room        Ethiopian Restaurant   Dumpling Restaurant   
2                      Café                 Yoga Studio    Mexican Restaurant   
3                  Boutique  Construction & Landscaping           Event Space   
4                       Bar                        Bank    Mexican Restaurant   

      10th Most Common Venue  
index                         
0        Dumpling Restaurant  
1           Department Store  
2          French Restaurant  
3       Ethiopian Restaurant  
4         Italian Restaurant

In [56]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough              Neighborhood   Latitude  \
index                                                                 
5             M9A    Etobicoke          Islington Avenue  43.667856   
45            M2L   North York  York Mills, Silver Hills  43.757490   
95            M1X  Scarborough               Upper Rouge  43.836125   

       Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                          
5     -79.532242             NaN                   NaN                   NaN   
45    -79.374714             NaN                   NaN                   NaN   
95    -79.205636             NaN                   NaN                   NaN   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
index                                                                     
5                       NaN                   NaN                   NaN   
45                      NaN                   NaN                   NaN   
95                      NaN                   NaN                   NaN   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
index                                                                     
5                       NaN                   NaN                   NaN   
45                      NaN                   NaN                   NaN   
95                      NaN                   NaN                   NaN   

      9th Most Common Venue 10th Most Common Venue  
index                                               
5                       NaN                    NaN  
45                      NaN                    NaN  
95                      NaN                    NaN

In [57]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Cluster 1

In [58]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
index                                                            
1           North York             0.0             Pizza Place   
8            East York             0.0             Pizza Place   
17           Etobicoke             0.0             Pizza Place   
19        East Toronto             0.0             Pizza Place   
50          North York             0.0  Furniture / Home Store   
51         Scarborough             0.0     American Restaurant   
56                York             0.0          Sandwich Place   
70           Etobicoke             0.0             Pizza Place   
72          North York             0.0                Pharmacy   
77           Etobicoke             0.0             Pizza Place   
79     Central Toronto             0.0             Pizza Place   
82         Scarborough             0.0             Pizza Place   
89           Etobicoke             0.0           Grocery Store   
90         Scarborough             0.0    Fast Food Restaurant   
93           Etobicoke             0.0             Pizza Place   

      2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
index                                                                      
1               Coffee Shop  Portuguese Restaurant          Hockey Arena   
8        Athletics & Sports              Gastropub          Intersection   
17               Beer Store                   Park           Coffee Shop   
19        Health Food Store           Neighborhood                 Trail   
50             Intersection            Pizza Place            Comic Shop   
51                    Motel    Dumpling Restaurant        Discount Store   
56     Fast Food Restaurant         Discount Store  Caribbean Restaurant   
70              Coffee Shop         Sandwich Place        Discount Store   
72              Coffee Shop            Supermarket               Butcher   
77           Sandwich Place               Bus Line            Donut Shop   
79           Sandwich Place           Dessert Shop           Coffee Shop   
82     Fast Food Restaurant               Pharmacy          Noodle House   
89              Pizza Place    Fried Chicken Joint              Pharmacy   
90              Pizza Place     Chinese Restaurant           Supermarket   
93             Skating Rink               Pharmacy           Coffee Shop   

      5th Most Common Venue        6th Most Common Venue  \
index                                                      
1         Electronics Store  Eastern European Restaurant   
8               Flea Market                     Pharmacy   
17                     Café               Shopping Plaza   
19                      Pub  Eastern European Restaurant   
50             Concert Hall           Falafel Restaurant   
51      Distribution Center                      Dog Run   
56               Donut Shop                        Diner   
70       Chinese Restaurant                 Intersection   
72              Pizza Place                  Escape Room   
77           Discount Store          Distribution Center   
79                      Gym           Italian Restaurant   
82       Italian Restaurant          Fried Chicken Joint   
89           Sandwich Place         Fast Food Restaurant   
90           Sandwich Place                         Bank   
93                      Pub               Sandwich Place   

             7th Most Common Venue        8th Most Common Venue  \
index                                                             
1                      Escape Room         Ethiopian Restaurant   
8                             Café                         Bank   
17                    Liquor Store                     Pharmacy   
19               Electronics Store          Dumpling Restaurant   
50                     Event Space         Ethiopian Restaurant   
51                Doner Restaurant                   Donut Shop   
56             Distribution Center        

### Cluster 2

In [59]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                                
2           Downtown Toronto             1.0           Coffee Shop   
3                 North York             1.0        Clothing Store   
4               Queen's Park             1.0           Coffee Shop   
7                 North York             1.0                   Gym   
9           Downtown Toronto             1.0        Clothing Store   
...                      ...             ...                   ...   
97          Downtown Toronto             1.0           Coffee Shop   
98                 Etobicoke             1.0                 River   
99          Downtown Toronto             1.0           Coffee Shop   
100    East Toronto Business             1.0    Light Rail Station   
102                Etobicoke             1.0         Grocery Store   

            2nd Most Common Venue  3rd Most Common Venue  \
index                                                      
2                            Park                 Bakery   
3               Accessories Store  Vietnamese Restaurant   
4                Sushi Restaurant            Yoga Studio   
7      Construction & Landscaping    Japanese Restaurant   
9                     Coffee Shop                   Café   
...                           ...                    ...   
97                           Café                  Hotel   
98                           Pool             Smoke Shop   
99               Sushi Restaurant    Japanese Restaurant   
100                 Auto Workshop                   Park   
102             Convenience Store          Tanning Salon   

      4th Most Common Venue   5th Most Common Venue  \
index                                                 
2                       Pub                 Theater   
3                 Gift Shop             Coffee Shop   
4       Japanese Restaurant                    Café   
7      Caribbean Restaurant                    Café   
9        Italian Restaurant         Bubble Tea Shop   
...                     ...                     ...   
97                      Gym              Restaurant   
98              Yoga Studio              Donut Shop   
99               Restaurant                    Park   
100              Comic Shop             Pizza Place   
102         Supplement Shop  Thrift / Vintage Store   

           6th Most Common Venue     7th Most Common Venue  \
index                                                        
2                 Breakfast Spot                      Café   
3         Furniture / Home Store                  Boutique   
4                          Diner                       Bar   
7           Ethiopian Restaurant               Escape Room   
9      Middle Eastern Restaurant            Cosmetics Shop   
...                          ...                       ...   
97           Japanese Restaurant          Asian Restaurant   
98                         Diner            Discount Store   
99                   Men's Store  Mediterranean Restaurant   
100             Recording Studio                Restaurant   
102               Sandwich Place             Burrito Place   

            8th Most Common Venue        9th Most Common Venue  \
index                                                            
2                     Yoga Studio           Mexican Restaurant   
3      Construction & Landscaping                  Event Space   
4                            Bank           Mexican Restaurant   
7               Electronics Store  Eastern European Restaurant   
9                     Pizza Place             Ramen Restaurant   
...                           ...                          ...   
97                  Deli / Bodega                  Salad Place   
98            Distribution Center                      Dog Run   
99                            Pub                  Yoga Studio   
100                 Burrito Place                   Skate Park   
102           

### Cluster 3

In [60]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                      
6      Scarborough             2.0  Fast Food Restaurant   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
6       Dumpling Restaurant        Discount Store   Distribution Center   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
6                   Dog Run      Doner Restaurant            Donut Shop   

      8th Most Common Venue        9th Most Common Venue  \
index                                                      
6                 Drugstore  Eastern European Restaurant   

      10th Most Common Venue  
index                         
6              Women's Store

### Cluster 4

In [61]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
index                                                                           
57     North York             3.0        Baseball Field           Yoga Studio   
101     Etobicoke             3.0                  Pool        Baseball Field   

      3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
index                                                                     
57      Dumpling Restaurant   Distribution Center               Dog Run   
101             Yoga Studio             Drugstore        Discount Store   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
index                                                                     
57         Doner Restaurant            Donut Shop             Drugstore   
101     Distribution Center               Dog Run      Doner Restaurant   

             9th Most Common Venue 10th Most Common Venue  
index                                                      
57     Eastern European Restaurant   Fast Food Restaurant  
101                     Donut Shop    Dumpling Restaurant

### Cluster 5

In [62]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
index                                                                 
0                  North York             4.0              Bus Stop   
21                       York             4.0                  Park   
35     East York/East Toronto             4.0                  Park   
40                 North York             4.0               Airport   
61            Central Toronto             4.0                  Park   
64                       York             4.0                  Park   
66                 North York             4.0                  Park   
85                Scarborough             4.0          Intersection   
91           Downtown Toronto             4.0                  Park   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
index                                                                     
0                      Park     Food & Drink Shop           Yoga Studio   
21            Women's Store                  Pool             Drugstore   
35            Metro Station     Convenience Store           Yoga Studio   
40         Business Service                  Park           Yoga Studio   
61                 Bus Line           Swim School           Yoga Studio   
64        Convenience Store         Jewelry Store           Yoga Studio   
66        Convenience Store           Yoga Studio   Dumpling Restaurant   
85               Playground                  Park             Drugstore   
91               Playground                 Trail           Yoga Studio   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
index                                                                     
0                 Drugstore   Distribution Center               Dog Run   
21                    Diner        Discount Store   Distribution Center   
35      Distribution Center               Dog Run      Doner Restaurant   
40      Dumpling Restaurant   Distribution Center               Dog Run   
61                Drugstore   Distribution Center               Dog Run   
64      Dumpling Restaurant   Distribution Center               Dog Run   
66           Discount Store   Distribution Center               Dog Run   
85                    Diner        Discount Store   Distribution Center   
91               Donut Shop                 Diner        Discount Store   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
index                                                                     
0          Doner Restaurant            Donut Shop    Dumpling Restaurant  
21                  Dog Run      Doner Restaurant             Donut Shop  
35               Donut Shop             Drugstore    Dumpling Restaurant  
40         Doner Restaurant            Donut Shop              Drugstore  
61         Doner Restaurant            Donut Shop    Dumpling Restaurant  
64         Doner Restaurant            Donut Shop              Drugstore  
66         Doner Restaurant            Donut Shop              Drugstore  
85                  Dog Run      Doner Restaurant             Donut Shop  
91      Distribution Center               Dog Run       Doner Restaurant